pkg/engine/variables/vars.go

```diff
@@ -171,7 +180,26 @@ func substituteReferencesIfAny(log logr.Logger) jsonUtils.Action {
 	})
 }
 
-func substituteVariablesIfAny(log logr.Logger, ctx context.EvalInterface) jsonUtils.Action {
+//VariableResolver defines the handler function for variable substitution
+type VariableResolver = func(ctx context.EvalInterface, variable string) (interface{}, error)
+
+// DefaultVariableResolver is used in all variable substitutions except preconditions
+func DefaultVariableResolver(ctx context.EvalInterface, variable string) (interface{}, error) {
+	return ctx.Query(variable)
+}
+
+// PreconditionsVariableResolver is used to substitute vars in preconditions.
+// It returns empty string if error occured during substitution
+func PreconditionsVariableResolver(ctx context.EvalInterface, variable string) (interface{}, error) {
+	value, err := DefaultVariableResolver(ctx, variable)
+	if err != nil || value == nil {
```

Comment:
Can we return something informational to indicate cases 2 and 3?

Reply:
Ping @kacejot.

Reply:
@realshuting, yes. There is new error type returned, if variable is not found. It is provided in the PR: https://github.com/kyverno/kyverno/pull/1907
It should be merged before this PR.